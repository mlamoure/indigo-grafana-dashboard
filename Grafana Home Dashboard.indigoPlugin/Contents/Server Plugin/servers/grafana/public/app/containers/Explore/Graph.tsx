import $ from 'jquery';
import React, { Component } from 'react';
import moment from 'moment';

import 'vendor/flot/jquery.flot';
import 'vendor/flot/jquery.flot.time';
import * as dateMath from 'app/core/utils/datemath';
import TimeSeries from 'app/core/time_series2';

import Legend from './Legend';

// Copied from graph.ts
function time_format(ticks, min, max) {
  if (min && max && ticks) {
    var range = max - min;
    var secPerTick = range / ticks / 1000;
    var oneDay = 86400000;
    var oneYear = 31536000000;

    if (secPerTick <= 45) {
      return '%H:%M:%S';
    }
    if (secPerTick <= 7200 || range <= oneDay) {
      return '%H:%M';
    }
    if (secPerTick <= 80000) {
      return '%m/%d %H:%M';
    }
    if (secPerTick <= 2419200 || range <= oneYear) {
      return '%m/%d';
    }
    return '%Y-%m';
  }

  return '%H:%M';
}

const FLOT_OPTIONS = {
  legend: {
    show: false,
  },
  series: {
    lines: {
      linewidth: 1,
      zero: false,
    },
    shadowSize: 0,
  },
  grid: {
    minBorderMargin: 0,
    markings: [],
    backgroundColor: null,
    borderWidth: 0,
    // hoverable: true,
    clickable: true,
    color: '#a1a1a1',
    margin: { left: 0, right: 0 },
    labelMarginX: 0,
  },
  // selection: {
  //   mode: 'x',
  //   color: '#666',
  // },
  // crosshair: {
  //   mode: 'x',
  // },
};

class Graph extends Component<any, any> {
  componentDidMount() {
    this.draw();
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.data !== this.props.data ||
      prevProps.options !== this.props.options ||
      prevProps.split !== this.props.split ||
      prevProps.height !== this.props.height
    ) {
      this.draw();
    }
  }

  draw() {
    const { data, options: userOptions } = this.props;
    if (!data) {
      return;
    }
    const series = data.map((ts: TimeSeries) => ({
      color: ts.color,
      label: ts.label,
      data: ts.getFlotPairs('null'),
    }));

    const $el = $(`#${this.props.id}`);
    const ticks = $el.width() / 100;
    let { from, to } = userOptions.range;
    if (!moment.isMoment(from)) {
      from = dateMath.parse(from, false);
    }
    if (!moment.isMoment(to)) {
      to = dateMath.parse(to, true);
    }
    const min = from.valueOf();
    const max = to.valueOf();
    const dynamicOptions = {
      xaxis: {
        mode: 'time',
        min: min,
        max: max,
        label: 'Datetime',
        ticks: ticks,
        timeformat: time_format(ticks, min, max),
      },
    };
    const options = {
      ...FLOT_OPTIONS,
      ...dynamicOptions,
      ...userOptions,
    };
    $.plot($el, series, options);
  }

  render() {
    const { data, height } = this.props;
    return (
      <div className="panel-container">
        <div id={this.props.id} className="explore-graph" style={{ height }} />
        <Legend data={data} />
      </div>
    );
  }
}

export default Graph;
