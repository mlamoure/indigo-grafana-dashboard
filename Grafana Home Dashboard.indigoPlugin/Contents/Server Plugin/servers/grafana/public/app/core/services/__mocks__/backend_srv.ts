const backendSrv = {
  get: jest.fn(),
  getDashboard: jest.fn(),
  getDashboardByUid: jest.fn(),
  getFolderByUid: jest.fn(),
  post: jest.fn(),
  resolveCancelerIfExists: jest.fn(),
};

export function getBackendSrv() {
  return backendSrv;
}
