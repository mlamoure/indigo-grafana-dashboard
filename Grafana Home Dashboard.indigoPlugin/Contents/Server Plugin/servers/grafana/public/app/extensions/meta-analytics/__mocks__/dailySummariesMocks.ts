import { DashboardDailySummaryDTO } from '../api';

export const getMockDailySummaries = (): DashboardDailySummaryDTO[] => {
  const dailySummaries = [];
  for (let i = 1; i <= 30; i++) {
    dailySummaries.push({
      day: `2020-05-${i < 10 ? '0' + i : i}`,
      dashboardId: 0,
      views: i * 28,
      queries: 0,
      errors: 0,
      loadDuration: 0,
    });
  }
  return dailySummaries;
};
