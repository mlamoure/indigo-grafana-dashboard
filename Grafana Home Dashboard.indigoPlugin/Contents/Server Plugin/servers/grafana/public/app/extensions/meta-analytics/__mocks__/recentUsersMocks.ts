import { UserViewDTO } from '../api';

export const getMockRecentUsers = (): UserViewDTO[] => {
  const recentUsers = [];

  for (let i = 1; i <= 10; i++) {
    recentUsers.push({
      user: {
        id: i,
        name: `User ${i}`,
        avatarUrl: '/avatar/c84258e9c39059a89ab77d846ddab909',
        login: `user${i}`,
        email: i % 2 === 0 ? `user${i}@localhost.com` : undefined,
        hasCustomAvatar: i % 2 === 0,
      },
      viewed: new Date(new Date().setDate(new Date().getDate() - (i + 1))).toISOString(),
    });
  }

  return recentUsers;
};
