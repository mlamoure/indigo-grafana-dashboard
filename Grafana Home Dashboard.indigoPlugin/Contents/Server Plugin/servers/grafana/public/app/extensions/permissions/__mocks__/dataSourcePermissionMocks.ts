import { DataSourcePermission } from '../../types';

export const getMockDataSourcePermissionsUser = (): DataSourcePermission => {
  return {
    created: '2018-10-10T16:50:45+02:00',
    datasourceId: 1,
    id: 2,
    permission: 1,
    permissionName: 'Query',
    updated: '2018-10-10T16:50:45+02:00',
    userAvatarUrl: '/avatar/926aa85c6bcefa0b4deca3223f337ae1',
    userEmail: 'test@test.com',
    userId: 3,
    userLogin: 'testUser',
  };
};

export const getMockDataSourcePermissionsTeam = (): DataSourcePermission => {
  return {
    created: '2018-10-10T16:57:09+02:00',
    datasourceId: 1,
    id: 6,
    permission: 1,
    permissionName: 'Query',
    team: 'A-team',
    teamAvatarUrl: '/avatar/93c0801b955cbd443a8cfa91a401d7bc',
    teamId: 1,
    updated: '2018-10-10T16:57:09+02:00',
  };
};
