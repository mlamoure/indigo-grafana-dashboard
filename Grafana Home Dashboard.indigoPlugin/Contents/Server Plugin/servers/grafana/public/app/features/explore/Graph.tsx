import $ from 'jquery';
import React, { PureComponent } from 'react';
import difference from 'lodash/difference';

import 'vendor/flot/jquery.flot';
import 'vendor/flot/jquery.flot.time';
import 'vendor/flot/jquery.flot.selection';
import 'vendor/flot/jquery.flot.stack';

import { GraphLegend, LegendItem, LegendDisplayMode } from '@grafana/ui';
import { TimeZone, AbsoluteTimeRange } from '@grafana/data';
import TimeSeries from 'app/core/time_series2';

const MAX_NUMBER_OF_TIME_SERIES = 20;

// Copied from graph.ts
function time_format(ticks, min, max) {
  if (min && max && ticks) {
    const range = max - min;
    const secPerTick = range / ticks / 1000;
    const oneDay = 86400000;
    const oneYear = 31536000000;

    if (secPerTick <= 45) {
      return '%H:%M:%S';
    }
    if (secPerTick <= 7200 || range <= oneDay) {
      return '%H:%M';
    }
    if (secPerTick <= 80000) {
      return '%m/%d %H:%M';
    }
    if (secPerTick <= 2419200 || range <= oneYear) {
      return '%m/%d';
    }
    return '%Y-%m';
  }

  return '%H:%M';
}

const FLOT_OPTIONS = {
  legend: {
    show: false,
  },
  series: {
    lines: {
      linewidth: 1,
      zero: false,
    },
    shadowSize: 0,
  },
  grid: {
    minBorderMargin: 0,
    markings: [],
    backgroundColor: null,
    borderWidth: 0,
    // hoverable: true,
    clickable: true,
    color: '#a1a1a1',
    margin: { left: 0, right: 0 },
    labelMarginX: 0,
  },
  selection: {
    mode: 'x',
    color: '#666',
  },
  // crosshair: {
  //   mode: 'x',
  // },
};

interface GraphProps {
  data: any[];
  height?: number;
  width?: number;
  id?: string;
  range: AbsoluteTimeRange;
  timeZone: TimeZone;
  split?: boolean;
  userOptions?: any;
  onChangeTime?: (range: AbsoluteTimeRange) => void;
  onToggleSeries?: (alias: string, hiddenSeries: string[]) => void;
}

interface GraphState {
  /**
   * Type parameter refers to the `alias` property of a `TimeSeries`.
   * Consequently, all series sharing the same alias will share visibility state.
   */
  hiddenSeries: string[];
  showAllTimeSeries: boolean;
}

export class Graph extends PureComponent<GraphProps, GraphState> {
  $el: any;
  dynamicOptions = null;

  state = {
    hiddenSeries: [],
    showAllTimeSeries: false,
  };

  getGraphData(): TimeSeries[] {
    const { data } = this.props;

    return this.state.showAllTimeSeries ? data : data.slice(0, MAX_NUMBER_OF_TIME_SERIES);
  }

  componentDidMount() {
    this.draw();
    this.$el = $(`#${this.props.id}`);
    this.$el.bind('plotselected', this.onPlotSelected);
  }

  componentDidUpdate(prevProps: GraphProps, prevState: GraphState) {
    if (
      prevProps.data !== this.props.data ||
      prevProps.range !== this.props.range ||
      prevProps.split !== this.props.split ||
      prevProps.height !== this.props.height ||
      prevProps.width !== this.props.width ||
      prevState.hiddenSeries !== this.state.hiddenSeries
    ) {
      this.draw();
    }
  }

  componentWillUnmount() {
    this.$el.unbind('plotselected', this.onPlotSelected);
  }

  onPlotSelected = (event: JQueryEventObject, ranges) => {
    const { onChangeTime } = this.props;
    if (onChangeTime) {
      this.props.onChangeTime({
        from: ranges.xaxis.from,
        to: ranges.xaxis.to,
      });
    }
  };

  getDynamicOptions() {
    const { range, width, timeZone } = this.props;
    const ticks = (width || 0) / 100;
    const min = range.from;
    const max = range.to;
    return {
      xaxis: {
        mode: 'time',
        min: min,
        max: max,
        label: 'Datetime',
        ticks: ticks,
        timezone: timeZone,
        timeformat: time_format(ticks, min, max),
      },
    };
  }

  onShowAllTimeSeries = () => {
    this.setState(
      {
        showAllTimeSeries: true,
      },
      this.draw
    );
  };

  draw() {
    const { userOptions = {} } = this.props;
    const { hiddenSeries } = this.state;
    const data = this.getGraphData();

    const $el = $(`#${this.props.id}`);
    let series = [{ data: [[0, 0]] }];

    if (data && data.length > 0) {
      series = data
        .filter((ts: TimeSeries) => hiddenSeries.indexOf(ts.alias) === -1)
        .map((ts: TimeSeries) => ({
          color: ts.color,
          label: ts.label,
          data: ts.getFlotPairs('null'),
        }));
    }

    this.dynamicOptions = this.getDynamicOptions();

    const options = {
      ...FLOT_OPTIONS,
      ...this.dynamicOptions,
      ...userOptions,
    };

    $.plot($el, series, options);
  }

  getLegendItems = (): LegendItem[] => {
    const { hiddenSeries } = this.state;
    const data = this.getGraphData();

    return data.map(series => {
      return {
        label: series.alias,
        color: series.color,
        isVisible: hiddenSeries.indexOf(series.alias) === -1,
        yAxis: 1,
      };
    });
  };

  onSeriesToggle(label: string, event: React.MouseEvent<HTMLElement>) {
    // This implementation is more or less a copy of GraphPanel's logic.
    // TODO: we need to use Graph's panel controller or split it into smaller
    // controllers to remove code duplication. Right now we cant easily use that, since Explore
    // is not using DataFrame for graph yet

    const exclusive = event.ctrlKey || event.metaKey || event.shiftKey;

    this.setState((state, props) => {
      const { data, onToggleSeries } = props;
      let nextHiddenSeries: string[] = [];
      if (exclusive) {
        // Toggling series with key makes the series itself to toggle
        if (state.hiddenSeries.indexOf(label) > -1) {
          nextHiddenSeries = state.hiddenSeries.filter(series => series !== label);
        } else {
          nextHiddenSeries = state.hiddenSeries.concat([label]);
        }
      } else {
        // Toggling series with out key toggles all the series but the clicked one
        const allSeriesLabels = data.map(series => series.label);

        if (state.hiddenSeries.length + 1 === allSeriesLabels.length) {
          nextHiddenSeries = [];
        } else {
          nextHiddenSeries = difference(allSeriesLabels, [label]);
        }
      }

      if (onToggleSeries) {
        onToggleSeries(label, nextHiddenSeries);
      }

      return {
        hiddenSeries: nextHiddenSeries,
      };
    });
  }

  render() {
    const { height = 100, id = 'graph' } = this.props;
    return (
      <>
        {this.props.data && this.props.data.length > MAX_NUMBER_OF_TIME_SERIES && !this.state.showAllTimeSeries && (
          <div className="time-series-disclaimer">
            <i className="fa fa-fw fa-warning disclaimer-icon" />
            {`Showing only ${MAX_NUMBER_OF_TIME_SERIES} time series. `}
            <span className="show-all-time-series" onClick={this.onShowAllTimeSeries}>{`Show all ${
              this.props.data.length
            }`}</span>
          </div>
        )}
        <div id={id} className="explore-graph" style={{ height }} />

        <GraphLegend
          items={this.getLegendItems()}
          displayMode={LegendDisplayMode.List}
          placement="under"
          onLabelClick={(item, event) => {
            this.onSeriesToggle(item.label, event);
          }}
        />
      </>
    );
  }
}

export default Graph;
