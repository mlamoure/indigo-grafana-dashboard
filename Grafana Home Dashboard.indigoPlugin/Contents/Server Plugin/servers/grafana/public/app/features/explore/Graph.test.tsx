import React from 'react';
import { shallow } from 'enzyme';
import Graph from './Graph';
import { mockData } from './__mocks__/mockData';

const setup = (propOverrides?: object) => {
  const props = Object.assign(
    {
      data: mockData().slice(0, 19),
      options: {
        interval: '20s',
        range: { from: 'now-6h', to: 'now' },
        targets: [
          {
            format: 'time_series',
            instant: false,
            hinting: true,
            expr: 'prometheus_http_request_duration_seconds_bucket',
          },
        ],
      },
    },
    propOverrides
  );

  // Enzyme.shallow did not work well with jquery.flop. Mocking the draw function.
  Graph.prototype.draw = jest.fn();

  const wrapper = shallow(<Graph {...props} />);
  const instance = wrapper.instance() as Graph;

  return {
    wrapper,
    instance,
  };
};

describe('Render', () => {
  it('should render component', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with disclaimer', () => {
    const { wrapper } = setup({
      data: mockData(),
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should show query return no time series', () => {
    const { wrapper } = setup({
      data: [],
    });

    expect(wrapper).toMatchSnapshot();
  });
});
