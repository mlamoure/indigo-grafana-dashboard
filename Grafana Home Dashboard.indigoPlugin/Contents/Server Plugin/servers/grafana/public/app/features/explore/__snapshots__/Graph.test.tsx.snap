// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<Fragment>
  <div
    className="explore-graph"
    id="graph"
    style={
      Object {
        "height": 100,
      }
    }
  />
  <GraphLegend
    displayMode="list"
    items={
      Array [
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
      ]
    }
    onLabelClick={[Function]}
    placement="under"
  />
</Fragment>
`;

exports[`Render should render component with disclaimer 1`] = `
<Fragment>
  <div
    className="time-series-disclaimer"
  >
    <i
      className="fa fa-fw fa-warning disclaimer-icon"
    />
    Showing only 20 time series. 
    <span
      className="show-all-time-series"
      onClick={[Function]}
    >
      Show all 27
    </span>
  </div>
  <div
    className="explore-graph"
    id="graph"
    style={
      Object {
        "height": 100,
      }
    }
  />
  <GraphLegend
    displayMode="list"
    items={
      Array [
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
        Object {
          "color": undefined,
          "isVisible": true,
          "label": undefined,
          "yAxis": 1,
        },
      ]
    }
    onLabelClick={[Function]}
    placement="under"
  />
</Fragment>
`;

exports[`Render should show query return no time series 1`] = `
<Fragment>
  <div
    className="explore-graph"
    id="graph"
    style={
      Object {
        "height": 100,
      }
    }
  />
  <GraphLegend
    displayMode="list"
    items={Array []}
    onLabelClick={[Function]}
    placement="under"
  />
</Fragment>
`;
